### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by mattiaslundberg**
Add integration tests, since it's kindof hard to test with a real environment locally. Trying to add a "mocked slack" that's simple to use.

Tests run fine locally but fails on CI, trying to figure out why but any pointers are appreciated!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -23,9 +24,21 @@ stages:
...
   - stage: Run tests
+    env:
+    - AWS_ACCESS_KEY_ID: dummy
+    - AWS_SECRET_ACCESS_KEY: dummy
+    - AWS_DEFAULT_REGION: eu-north-1
     script:
     - pipenv install --dev
-    - pipenv run pytest -v
+    - pipenv run pytest -v -m "not integration"
+    - git clone https://github.com/codelabsab/timereport-api.git ../timereport-api
```
Comment (parberge) [inline .travis.yml:34]: Should we always run integration tests for all commits on all branches?

Comment (mattiaslundberg) [reply .travis.yml:34]: Good question, I think we should!
We still have fast tests (full run takes < 3 minutes on travis) and I always prefer fast feedback if I make a mistake. Preferably feedback on local tests (I get this here) but at least feedback from CI before my PR is reviewed. If we decide that this is too slow I think we should start using branch filters, for example not running integration tests for branches starting with `wip/` (or something similar) or not running integration tests for commits with special tag in the message (think `[skip ci]` for integration tests only).

Does that make sense?

Comment (parberge) [reply .travis.yml:34]: Sure, let's go with this for now and evaluate 👍

File: tests/utils.py [hunk 2]
```diff
@@ -37,4 +58,9 @@ def call_from_slack(
             body=body,
         )
 
-    return dict(slack_message=m.call_args[1]["message"], response=response)
+    return dict(slack_message=mock_request.post.call_args, response=response)
+
+
+def get_raw_block_text(slack_message):
+    blocks = slack_message[1]["json"]["blocks"]
+    return "".join([b.get("text", {}).get("text", "") for b in blocks])
```
Comment (parberge) [inline tests/utils.py:66]: Having a hard time understanding this one. Might it be rewritten to make it more obvious what it does?

File: tests/test_app.py [hunk 3]
```diff
@@ -1,8 +1,95 @@
-from tests.utils import call_from_slack
+import random
+
+import pytest
+from tests.utils import call_from_slack, get_raw_block_text, respond_interactively
 
 
 def test_help_command(chalice_app):
-    r = call_from_slack(chalice_app, ["help"], "1234", "mattias")
+    r = call_from_slack(chalice_app, "help", f"{random.randint(0, 10000)}", "mattias")
```
Comment (parberge) [inline tests/test_app.py:8]: I would like to use the keyword arg style to make it more clear on what the argument is used for.
```suggestion
    r = call_from_slack(chalice_app=chalice_app, full_command="help", user_id=f"{random.randint(0, 10000)}", user_name="mattias")
```

There are more of these to edit if we agree on this.

File: tests/test_app.py [hunk 4]
```diff
@@ -1,8 +1,95 @@
...
+def test_add_command_accepted(chalice_app):
+    user_id = f"{random.randint(0, 10000)}"
+    r = call_from_slack(chalice_app, "add vab today 8", user_id, "mattias")
+
+    assert r["response"]["statusCode"] == 200
+    assert r["slack_message"][1]["json"]["text"] == "From timereport"
+    attachments = r["slack_message"][1]["json"]["attachments"]
+    assert attachments is not None
+    attachment = attachments[0]
+    assert len(attachment["actions"]) > 0
```
Comment (parberge) [inline tests/test_app.py:34]:
```suggestion
    assert attachment["actions"]
```
Empty list is consider falsy in python

**Comment by mattiaslundberg** [comment]
> I see that we check for specific strings in the output of our tests.

> Do we really want to be that specific? If we change the output text by just adding an emoji or whatever the tests will then begin to fail.

> I think it might be sufficient to just check that the data type was correct.

I've tried to partially match for keywords at most places, will look it over and change to try to make it more robust. Usually not a problem to update the test as well when changing a text (but can take some time figuring out why the test failed if not obvious). I don't think checking just output type is enough since there isn't any structured response (http response and different slack messages combined).

I agree with the other comments and will try to fix them :)

**Pull request merged**